core/src/main/java/org/elasticsearch/search/slice/SliceBuilder.java

```diff
@@ -181,14 +189,54 @@ public int hashCode() {
         return Objects.hash(this.field, this.id, this.max);
     }
 
-    public Query toFilter(QueryShardContext context) {
+    public Query toFilter(QueryShardContext context, int shardId, int numShards) {
         final MappedFieldType type = context.fieldMapper(field);
         if (type == null) {
             throw new IllegalArgumentException("field " + field + " not found");
         }
+
         if (UidFieldMapper.NAME.equals(field)) {
-            return new TermsSliceQuery(field, id, max);
+            if (numShards == 1) {
+                return new TermsSliceQuery(field, id, max);
+            }
+
+            if (max >= numShards) {
```

Comment:
I think we should do this even if we use docvlaues? I think we should have consistent slicing no matter how it's done!